Parser generation forecast data JP

alejandranavcas opened the pull request:
## Description

This PR is to include parser for generation forecast data in all zones in Japan (JP).


### Double check

- [x] I have tested my parser changes locally with `poetry run test_parser "zone_key"`
- [x] I have run `pnpx prettier@2 --write .` and `poetry run format` in the top level directory to format my changes.

VIKTORVAV99 commented on parsers/JP.py:
```diff
@@ -352,9 +361,178 @@ def parse_dt(row):
+                with z.open(target_file) as f:
+                    df = pd.read_csv(f, encoding="shift_jis", skiprows=startrow)
+            else:
+                print(f"File '{target_file}' not found in the zip archive.")
+                print(f"Available files: {z.namelist()}")
```
This should be a logger if you intend to use this for debugging in production.

VIKTORVAV99 commented on parsers/JP.py:
```diff
@@ -352,9 +361,178 @@ def parse_dt(row):
+
+    # Extract the correspondant information from the files for each zone
+    if zone_key == "JP-HKD" or zone_key == "JP-HR":
+        df = df[["日付", "時間帯_自", "エリア総発電量(kWh)"]]
```
Since the df is created in a conditional above it's potentially unbound here and downstream.

VIKTORVAV99 commented on parsers/JP.py:
```diff
@@ -352,9 +361,178 @@ def parse_dt(row):
+                f"{time_int:04d}" if time_int >= 100 else f"00{time_int:02d}"
+            )
+            time_obj = datetime.strptime(time_str_padded, "%H%M").time()
+            date_obj = datetime.strptime(str(int(event["Date"])), "%Y%m%d").date()
```
Why is this converted first to a int and then a string?

VIKTORVAV99 replied:
But the CSVs don't have any decimals and are stored as INTs directly, so the casting INTs should not be needed, so if there are decimals added that signals a bigger issue somewhere. Probably because we are using Pandas to read the CSV.

alejandranavcas replied:
Yes, that's true. I'll look into that then :)

alejandranavcas replied:
I found the issue: when using iterrows(), pandas converts the values to float because of the way it handles data internally. Now, I'll use itertuples() instead

VIKTORVAV99 reviewed (commented):
Looks like it works but I'm not a huge fan of casting converting these objects to various types instead of parsing them directly, what is the reason for this?

alejandranavcas commented:
Thanks for all the feedback! 
I've now implemented the changes you suggested and also added `generationForecast: JP.fetch_generation_forecast` in the parsers object in the config files. Let me know if more changes should be done!

VIKTORVAV99 commented on parsers/JP.py:
```diff
@@ -352,9 +361,195 @@ def parse_dt(row):
+    generation_list = TotalProductionList(logger)
+
+    for row in all_generation_events.itertuples():
+        if (
+            zone_key == "JP-HR"
+            or zone_key == "JP-ON"
+            or zone_key == "JP-TH"
+            or zone_key == "JP-CB"
+        ):
+            year, month, day = map(int, row.Date.split("/"))
+            hour, minute = map(int, row.Time.split(":"))
+            event_datetime = datetime(year, month, day, hour, minute).replace(
+                tzinfo=ZONE_INFO
+            )
+
+        elif zone_key == "JP-TK":
+            time_str_padded = (
+                f"{row.Time:04d}" if row.Time >= 100 else f"00{row.Time:02d}"
+            )
+            time_obj = time(
+                hour=int(time_str_padded[:2]), minute=int(time_str_padded[2:])
+            )
+            date_obj = date(
+                year=row.Date // 10000,
+                month=(row.Date % 10000) // 100,
+                day=row.Date % 100,
+            )
+            event_datetime = datetime.combine(date_obj, time_obj).replace(
+                tzinfo=ZONE_INFO
+            )
+        else:
+            date_int = int(row.Date)
+            year = date_int // 10000
+            month = (date_int % 10000) // 100
+            day = date_int % 100
+            hour, minute = map(int, row.Time.split(":"))
+            event_datetime = datetime(year, month, day, hour, minute).replace(
+                tzinfo=ZONE_INFO
+            )
```
```suggestion
        if (
            zone_key == "JP-HR"
            or zone_key == "JP-ON"
            or zone_key == "JP-TH"
            or zone_key == "JP-CB"
        ):
            event_datetime = datetime.strptime(
                f"{row.Date} {row.Time}", "%Y/%m/%d %H:%M"
            ).replace(tzinfo=ZONE_INFO)

        elif zone_key == "JP-TK":
            # Format as string in the format "YYYYMMDD HHMM" and parse
            event_datetime = datetime.strptime(
                f"{row.Date:08d} {row.Time:04d}", "%Y%m%d %H%M"
            ).replace(tzinfo=ZONE_INFO)
        else:
            # Use strptime to directly parse the date integer and time string
            event_datetime = datetime.strptime(
                f"{row.Date:08d} {row.Time}", "%Y%m%d %H:%M"
            ).replace(tzinfo=ZONE_INFO)
```
How about we do this instead? It gets rid of a lot of type issues and are less code overall.

alejandranavcas replied:
Yes, thank you 👍  that looks so much better! But there was a zone that was not working correctly (JP-KY), so I just included it in another elif condition separately. I will push the commit now

VIKTORVAV99 reviewed (approved):
LGTM! 🎉
